src/protocolsupportpocketstuff/hacks/playerheads/PlayerHeadsPacketListener.java

```diff
@@ -0,0 +1,617 @@
+			serializer.writeBytes(MiscSerializer.readBytes(serializer, serializer.readableBytes()));
+			event.setData(serializer);
+			new Thread() {
+				public void run() {
+					if (Skins.INSTANCE.hasPeSkin(url)) {
+						sendSkullSkin(Skins.INSTANCE.getPeSkin(url), con.getPlayer().getUniqueId(), true);
+					} else {
+						try {
+							BufferedImage image = ImageIO.read(new URL(url));
+							byte[] data = toData(image);
+							sendSkullSkin(data, con.getPlayer().getUniqueId(), true);
+							Skins.INSTANCE.cachePeSkin(url, data);
+						} catch (IOException e) {
+							e.printStackTrace();
+						}
+					}
+				}
+			}.start();
+			return;
+		}
+		if (packetId == PEPacketIDs.MOB_ARMOR_EQUIPMENT) {
+			long entityId = VarNumberSerializer.readVarLong(data);
+
+			if (!entityIdToUuid.containsKey(entityId))
+				return;
+
+			ItemStackWrapper itemStack = ItemStackSerializer.readItemStack(data, con.getVersion(), I18NData.DEFAULT_LOCALE, false); // Yes, it is from the client, but we don't want it to remap our stuff
+
+			if (itemStack.isNull() || itemStack.getTypeId() != SKULL_ITEM_ID) {
+				if (hasCustomSkull.contains(entityId)) {
+					hasCustomSkull.remove(entityId);
+					sendOriginalSkin(entityIdToUuid.get(entityId), false);
+				}
+				return;
+			}
+
+			NBTTagCompoundWrapper tag = itemStack.getTag();
+
+			String url = getUrlFromSkull(tag, true);
+
+			hasCustomSkull.add(entityId);
+			new Thread() {
+				public void run() {
+					UUID uuid = entityIdToUuid.get(entityId);
+					if (Skins.INSTANCE.hasPeSkin(url)) {
+						sendSkullSkin(Skins.INSTANCE.getPeSkin(url), uuid, false);
+					} else {
+						try {
+							BufferedImage image = ImageIO.read(new URL(url));
+							byte[] data = toData(image);
+							sendSkullSkin(data, uuid, false);
+							Skins.INSTANCE.cachePeSkin(url, data);
+						} catch (IOException e) {
+							e.printStackTrace();
+						}
+					}
+				}
+			}.start();
+			return;
+		}
+		if (packetId == PEPacketIDs.TILE_DATA_UPDATE) {
+			Position position = PositionSerializer.readPEPosition(data);
+			try {
+				NBTTagCompoundWrapper tag = NBTTagCompoundSerializer.readPeTag(data, true);
+
+				if (!isSkull(tag))
+					return;
+
+				handleSkull(position, tag, -1);
+
+				tag.remove("Owner"); // remove the owner tag, if we don't remove it, the skull stays with the default parameters (skeleton with rot 0)
+				event.setData(new TileDataUpdatePacket(position.getX(), position.getY(), position.getZ(), tag).encode(con));
+			} catch (Exception e) {
+				e.printStackTrace();
+			}
+			return;
+		}
+		if (packetId == PEPacketIDs.UPDATE_BLOCK) {
+			Position position = PositionSerializer.readPEPosition(data);
+			int id = VarNumberSerializer.readVarInt(data);
+			int flagsAndDataValue = VarNumberSerializer.readVarInt(data);
+
+			long asLong = position.asLong();
+
+			int flags = 0;
+
+			if ((flagsAndDataValue & 0x01) != 0) {
```

Comment:
Instead of all this: flags = AndDataValue & 0xF

Reply:
Oh maybe not. Leave it ill take a look. You might be able to do this more simple